Display "Issues to Resolve" in Product Feed

ecgan opened the pull request:
### Changes proposed in this Pull Request:

Closes woocommerce/woocommerce-admin#245 .

This PR calls the issues API (introduced in PR https://github.com/woocommerce/google-listings-and-ads/pull/421) and display the issues in the "Issues to Resolve" table.

Account type issue will be displayed in a red warning icon. Product type issue will be displayed in a yellow warning icon.

#### Technical note: 

In the `IssuesTableCard` component, I am essentially re-implementing my own `TableCard` component without showing the "per page" picker, because: 

- `TableCard` component uses `Pagination` component which has hardcoded page size options: 25, 50, 75, 100. For our use case, we want to have a smaller table in the page, and `Pagination` page size cannot be customized. See the [related code in `Pagination`](https://github.com/woocommerce/woocommerce-admin/blob/dde6b184ed6d14985402dc76f7078bf03b7985d7/packages/components/src/pagination/index.js#L12). (I've created issue woocommerce/woocommerce#32256.)
- `Pagination` component has a `showPerPagePicker` props that allows you to hide the "per page" picker, but it is not exposed by `TableCard` component. See the [related code in `TableCard`](https://github.com/woocommerce/woocommerce-admin/blob/dde6b184ed6d14985402dc76f7078bf03b7985d7/packages/components/src/table/index.js#L246-L253). (I've created issue woocommerce/woocommerce#32255.)

### Screenshots:

Loading:

![issues table - loading](https://user-images.githubusercontent.com/417342/115969362-296e9d00-a56f-11eb-98c6-0c10cf509050.png)

Error state:

![issues table - error](https://user-images.githubusercontent.com/417342/115969366-2ffd1480-a56f-11eb-8862-2a5f9faf78a4.png)

When there is no data: 

![issues table - no data](https://user-images.githubusercontent.com/417342/115969363-2bd0f700-a56f-11eb-8be7-cc32e25ff27e.png)

The text "No data to display" comes by default from the TableCard component. See: https://github.com/woocommerce/woocommerce-admin/blob/dde6b184ed6d14985402dc76f7078bf03b7985d7/packages/components/src/table/table.js#L329-L332

With data:

![issues table - with data](https://user-images.githubusercontent.com/417342/115969364-2d9aba80-a56f-11eb-8da4-51c70ada0b89.png)

### Detailed test instructions:

1. Open https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fproduct-feed
2. The "Issues to Resolve" table should display as per the above screenshots.
3. For product issue, there should be an Edit button to redirect to the product edit page.

### Changelog Note:

Call issues API and display the issues in the "Issues to Resolve" table.

ecgan committed:
Do custom table card implementation to support custom pagination.

We want only 5 rows per page in the table but the WooCommerce TableCard has hard-coded pagination options and does not support 5 per page.

ecgan committed:
Add @wordpress/url as dependency in package.json.

ecgan committed:
Add red filled error icon.

ecgan committed:
Display dummy issues in IssuesTableCard.

ecgan committed:
wp-data for getting MC issues.

ecgan committed:
Remove dummy data and use wp-data in useMCIssues hook.

ecgan committed:
Display MC Issues in IssuesTableCard.

ecgan committed:
Move key generation into reducer instead of action creator.

This is because action creator should not have knowledge of the state tree shape.

ecgan committed:
Show EditProductLink only when issue type is product.

ecgan committed:
Better CSS UI layout for ProductFeed.

ecgan committed:
Normalize issues data in wp-data store.

ecgan committed:
Simplify mc_issues reducer code.

ecgan commented:
@j111q ,

> 1. Can we have Sentence case rather than Title Case on all the headings/columns? eg. `Issues to resolve` and `Affected product` and `Suggested action`.

Sure, I'll work on that in this PR. (Update: done in https://github.com/woocommerce/google-listings-and-ads/pull/518/commits/f0c408c920a597c911cc9fffa7d36e88218a274d)

> 2. Can we add a 16px space between the table title and the question mark icon?

Yeah, it happens in quite a few places, I'll create a separate issue and address them together. (Update: see issue https://github.com/woocommerce/google-listings-and-ads/issues/529)

> 3. Can a user click on the column heading to "Sort by" that column heading? Eg. click on `Affected product` to sort by product title?

For this "Issues to resolve" table, no, I have intentionally disabled sorting here because the API does not have sorting available.

The API is developed in PR https://github.com/woocommerce/google-listings-and-ads/pull/421 by @layoutd .

> If no, then what determines the order of how are the items listed?

> By default, it should be listed with `Disapprovals` on top, and `Warnings` below; and subsequently within Disapprovals / Warnings, it should be grouped by `Affected product`.

So from PR woocommerce/woocommerce-admin#421: 

- Account-level issues will be listed first (In the UI, these will be displayed with red warning icon).
- Product-level issues are grouped by description (e.g. `Missing value [color]`) and sorted by WooCommerce product ID. This essentially means we would see all issues for Product A, and then all issues for Product B, etc. (In the UI, these will be displayed with yellow warning icon).

(Note: not sure if this would help you, in the PR woocommerce/woocommerce-admin#421 description, there is an "example REST response" that is collapse by default, you can click on it and see the sample API response data.)

@layoutd , tagging you here too in case you have something to add on.

ecgan committed:
Merge branch 'feature/product-statistics-with-api' into feature/product-feed-issues-call-api

ecgan committed:
Use sentence case.

ecgan committed:
Make action text link to Google documentation.

ecgan committed:
Merge branch 'feature/product-statistics-with-api' into feature/product-feed-issues-call-api

Conflicts:
	js/src/product-feed/index.js
	package.json

ecgan committed:
Fix error message in getMCIssues.

ecgan committed:
Simplify with useAppSelectDispatch, remove useMCIssues.

ecgan committed:
Merge branch 'feature/product-statistics-with-api' into feature/product-feed-issues-call-api

ecgan committed:
Add rememo package.

ecgan committed:
Use rememo createSelector in getMCIssues.

eason9487 reviewed (approved):
Testing well and LGTM. Left a comment for polishing and good to go.

j111q commented:
>So from PR woocommerce/woocommerce-admin#421:
> 
> Account-level issues will be listed first (In the UI, these will be displayed with red warning icon).
> Product-level issues are grouped by description (e.g. Missing value [color]) and sorted by WooCommerce product ID. This essentially means we would see all issues for Product A, and then all issues for Product B, etc. (In the UI, these will be displayed with yellow warning icon).

Got it, thanks for the clarification. 

To be very clear just in case - product-level issues also can have disapprovals (with the red disapproval icon). This is just aligning with Google Merchant Center's existing behaviour, and I would assume, the API they provide.
See their [documentation here under Issue Prioritization](https://support.google.com/merchants/answer/1680021?hl=en):
![image](https://user-images.githubusercontent.com/65582817/116856376-a5cb4500-ac2d-11eb-9070-7acf9c1be08f.png)


I will also add "sorting" as a potential future area of improvement.

ecgan committed:
Merge branch 'trunk' into feature/product-feed-issues-call-api

Conflicts:
	package-lock.json

ecgan committed:
Remove unused commented code.

ecgan committed:
Remove unneeded useMemo.

This is because we now have useRef and isEqual check inside useAppSelectDispatch.
